core/src/main/scala/org/http4s/EntityDecoder.scala

```diff
@@ -99,8 +99,12 @@ object EntityDecoder extends EntityDecoderInstances {
 
   private class OrDec[T](a: EntityDecoder[T], b: EntityDecoder[T]) extends EntityDecoder[T] {
     override def decode(msg: Message): DecodeResult[T] = {
-      if (a.matchesMediaType(msg)) a.decode(msg)
-      else b.decode(msg)
+      msg.headers.get(`Content-Type`).map { contentType =>
+        if (a.matchesMediaType(contentType.mediaType)) a.decode(msg)
+        else b.decode(msg)
+      }.getOrElse {
+        a.decode(msg).orElse(b.decode(msg))
```

Comment:
Its not necessarily a scalaz-stream issue: a stream can be effectual or not, its really up to the user. Http4s uses an effectful stream because otherwise we need to buffer the whole body which is a no-go in the general case. There has been talk of a middleware that would 'memorize' the body but it hasn't happened and I'm not sure it would be super useful anyway.